### ISSUE TITLE
Example references under keys ending with items or schema are not correctly resolved

### ISSUE THREAD

**Issue by allenhsu** [swagger-api/swagger-js#1801]
### Q&A (please complete the following information)
 - OS: macOS
 - Environment: Chrome 86.0.4240.111
 - Method of installation: npm
 - Swagger-Client version: 3.12.0
 - Swagger/OpenAPI version: both Swagger 2.0 and OpenAPI 3.0

### Content & configuration

We are using examples for some responses with `$ref`, however when the example is under a key ending with `items`, it won't be correctly absolutified and resolved. After checking the source code, I believe keys ending with `schema` should have the same issue.

Swagger/OpenAPI definition:

A simplified version of the definition is as following:

```yaml
responses:
    "200":
        description: "Success"
        content:
            application/json:
                schema:
                    type: object
                    properties:
                        cart_items:
                            type: object
                            additionalProperties:
                                $ref: "#/components/schemas/CartItem"
                            example:
                                $ref: "#/components/examples/CartItemExample"
                        cart_products:
                            type: object
                            additionalProperties:
                                $ref: "#/components/schemas/CartItem"
                            example:
                                $ref: "#/components/examples/CartItemExample"
```

Swagger-Client usage:

I'm using the raw SwaggerUI.

### Describe the bug you're encountering

![image](https://user-images.githubusercontent.com/368504/98213019-6faf5000-1f7f-11eb-9c04-a02751917683.png)

As you can see from the screenshot, the `cart_products` is correctly absolutified and expanded, but `cart_items` is not, while their definition are all the same except the key name.

### To reproduce...

1. Define the API as mentioned above, and inspect the API in SwaggerUI or other client.

### Expected behavior
I'm not 100% sure whether this usage of `$ref` is correct or not. But I'm expecting the behavior to be consistent, irrelevant to the key name.

### Screenshots

![image](https://user-images.githubusercontent.com/368504/98213019-6faf5000-1f7f-11eb-9c04-a02751917683.png)

### Additional context or thoughts

I've identified the root cause for the problem. It's due to this [`isFreelyNamed` method defined in `specmap/helpers.js`](https://github.com/swagger-api/swagger-js/blob/master/src/specmap/helpers.js#L46)

```
const freelyNamedAncestors = ['schema/example', 'items/example'];
freelyNamedAncestors.some((el) => parentStr.indexOf(el) > -1)
```

Because for the case mentioned above, the `parentStr` is something like `"paths//URL/get/responses/200/content/application/json/schema/properties/cart_items/example"`, which contains `items/example`, so the check returns `true`.

Then [this line in `specmap/lib/refs.js`](https://github.com/swagger-api/swagger-js/blob/master/src/specmap/lib/refs.js#L69) will skip resolution.

I think the quickest fix is:
1. append a ending `/` to `parentStr`
2. then check for `['/schema/example/', '/items/example/']` instead of `['schema/example', 'items/example']`

While the most robust solution should be splitting the path into path components and checking, instead of text matching.

**Comment by allenhsu** [comment]
Before I make changes and create a PR, it would be helpful if someone can explain what the check is for, and which cases should be matched, which cases should not be matched. For example: `paths://a/b/c/items/example` and `paths://a/b/c/items/example/d/e`, should both cases be skipped?

**Comment by hkosova** [comment]
> I'm not 100% sure whether this usage of $ref is correct or not. 

It is not. The `example` keyword does not support `$ref`.

The $refs in your definition are not supposed to be resolved. Those examples should be parsed as just objects with a property named `$ref` whose value is the literal string "#/components/examples/CartItemExample".

**Comment by allenhsu** [comment]
> > I'm not 100% sure whether this usage of $ref is correct or not.
> 
> It is not. The `example` keyword does not support `$ref`.
> 
> The $refs in your definition are not supposed to be resolved. Those examples should be parsed as just objects with a property named `$ref` whose value is the literal string "#/components/examples/CartItemExample".

Thanks for the explanation. Then I think what I got is just an unwanted side effect. However I still hope the behavior could be more consistent, rather than changing depending on the key name. On the other hand, is it possible to officially have support for this kind of usage of `$ref` in `example`?

**Comment by char0n** [comment]
Hi @allenhsu,

The behavior placement and behavior of `$ref` is defined either by OpenApi specification and JSON Schema specification respectively. I admit it can be pretty difficult to understand where `$ref` can really be used so let's break it down a bit.

Let's say we we talk about [OpenApi 3.0.3](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md) version. This specification has a dedicated [Reference Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#referenceObject). This Reference Object can only be used in places where OpenApi 3.0.3 specification explicitly says it can be used, for example in [Components Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#components-object).

Then OpenApi 3.0.3 further [states](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#data-types) that it's Data Types are defined in [JSON Schema Specification Wright Draft 00](https://tools.ietf.org/html/draft-wright-json-schema-00). OpenApi 3.0.3 specification has a special [Schema Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#schemaObject) to define a Data Type. Every Schema Object is subsequently defined as a subset of [JSON Schema Specification Wright Draft 00](https://tools.ietf.org/html/draft-wright-json-schema-00), which technically makes is a foreign specification embedded inside the OpenApi 3.0.3 specification. Now [JSON Schema Specification Wright Draft 00](https://tools.ietf.org/html/draft-wright-json-schema-00) specifically limits use of `$ref` to only wherever another schema is expected ([Section 7](https://tools.ietf.org/html/draft-wright-json-schema-00#section-7) specifically). This means that if `$ref` is used a property name it has no other semantic meaning, except being the name of the property.

To sum it up in layman's terms: the behavior how `$ref` is treated is consistently and is precisely defined. There is no room for different interpretation. The problem is that [swagger-js](https://github.com/swagger-api/swagger-js) is just an implementation of the specification and as with every implementation we have bugs and inconsistent. Though we're working hard on [rectifying those inconsistencies](https://github.com/swagger-api/swagger-js/pull/1810).

**Comment by allenhsu** [comment]
Hi @char0n,

Thanks for the response, especially the link to the PR. It makes a lot of sense to me to disable it since it's not defined in the spec.

Separately, I'm not sure if this [`isFreelyNamed`](https://github.com/swagger-api/swagger-js/blob/master/src/specmap/helpers.js#L46) method will still be in use after the changes, but its implementation is imperfect because it will yield false positive results.

**Comment by char0n** [comment]
@allenhsu,

No problem. As for `isFreelyNamed`, it cannot be removed at this time as it still handles cases that `skipResolutionTestFns` doesn't cover. IMHO this function can be replaced in future by `skipResolutionTestFns` completely.
